From 4e876982ea68f7cec44eefc35951d8926cfadf18 Mon Sep 17 00:00:00 2001
From: Luca Magrone <luca@magrone.cc>
Date: Fri, 18 Oct 2024 02:41:04 +0200
Subject: [PATCH] CIEID: Allow the user to exit with ctrl+q

Attach a AWTEventListener to the Toolkit that checks if the user presses
the combinatrion CTRL+Q. On press terminate the JVM.

Signed-off-by: Luca Magrone <luca@magrone.cc>
---
 CIEID/src/it/ipzs/cieid/MainApplication.java | 28 ++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/CIEID/src/it/ipzs/cieid/MainApplication.java b/CIEID/src/it/ipzs/cieid/MainApplication.java
index 79904d5..58b077b 100644
--- a/CIEID/src/it/ipzs/cieid/MainApplication.java
+++ b/CIEID/src/it/ipzs/cieid/MainApplication.java
@@ -1,11 +1,17 @@
 package it.ipzs.cieid;
 
 import java.awt.EventQueue;
+import java.awt.AWTEvent;
+import java.awt.Toolkit;
+import java.awt.Window;
+import java.awt.event.AWTEventListener;
+import java.awt.event.KeyEvent;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 
 import javax.swing.JFrame;
+import javax.swing.FocusManager;
 
 import ch.swingfx.twinkle.NotificationBuilder;
 import ch.swingfx.twinkle.event.NotificationEvent;
@@ -26,6 +32,7 @@ public class MainApplication {
 	public static void main(final String[] args) {
 		EventQueue.invokeLater(new Runnable() {
 			public void run() {
+				installKeyboardMonitor();
 
 				if(args.length > 0 && args[0].equals("pinwrong"))
 				{
@@ -51,6 +58,27 @@ public class MainApplication {
 		});
 	}
 
+	/**
+	 * Close the application on ctrl+q
+	 */
+	public static void installKeyboardMonitor() {
+		Toolkit.getDefaultToolkit().addAWTEventListener(new AWTEventListener() {
+			@Override
+			public void eventDispatched(AWTEvent event)
+			{
+				KeyEvent ke = (KeyEvent) event;
+				if (ke.getID() == KeyEvent.KEY_PRESSED)
+				{
+					if (ke.getKeyCode() == KeyEvent.VK_Q)
+					{
+						if (ke.isControlDown())
+							System.exit(0);
+					}
+				}
+			}
+		}, AWTEvent.KEY_EVENT_MASK);
+	}
+
 	public static void showUI(String[] args)
 	{
 		MainApplication window = new MainApplication(args);
-- 
2.43.5

